recipes/qt/all/conanfile.py

```diff
@@ -0,0 +1,687 @@
+
+        value = _getenvpath('CC')
+        if value:
+            args += ['QMAKE_CC="' + value + '"',
+                     'QMAKE_LINK_C="' + value + '"',
+                     'QMAKE_LINK_C_SHLIB="' + value + '"']
+
+        value = _getenvpath('CXX')
+        if value:
+            args += ['QMAKE_CXX="' + value + '"',
+                     'QMAKE_LINK="' + value + '"',
+                     'QMAKE_LINK_SHLIB="' + value + '"']
+
+        if tools.os_info.is_linux and self.settings.compiler == "clang":
+            args += ['QMAKE_CXXFLAGS+="-ftemplate-depth=1024"']
+
+        if self.options.qtwebengine and self.settings.os == "Linux":
+            args += ['-qt-webengine-ffmpeg',
+                     '-system-webengine-opus']
+
+        if self.options.config:
+            args.append(str(self.options.config))
+
+        os.mkdir('build_folder')
+        with tools.chdir('build_folder'):
+            with tools.vcvars(self.settings) if self.settings.compiler == "Visual Studio" else tools.no_op():
+                build_env = {"MAKEFLAGS": "j%d" % tools.cpu_count(), "PKG_CONFIG_PATH": [self.build_folder]}
+                if self.settings.os == "Windows":
+                    build_env["PATH"] = [os.path.join(self.source_folder, "qt5", "gnuwin32", "bin")]
+                with tools.environment_append(build_env):
+                    self.run("%s/qt5/configure %s" % (self.source_folder, " ".join(args)), run_environment=True)
+                    if tools.os_info.is_macos:
+                        with open("bash_env", "w") as f:
+                            f.write('export DYLD_LIBRARY_PATH="%s"' % ":".join(RunEnvironment(self).vars["DYLD_LIBRARY_PATH"]))
+                    with tools.environment_append({
+                        "BASH_ENV": os.path.abspath("bash_env")
+                        }) if tools.os_info.is_macos else tools.no_op():
+                        self.run(self._make_program(), run_environment=True)
+
+    def package(self):
+        with tools.chdir('build_folder'):
+            self.run("%s install" % self._make_program())
+        with open(os.path.join(self.package_folder, "bin", "qt.conf"), 'w') as f:
+            f.write("""[Paths]
+Prefix = ..
+ArchData = bin/archdatadir
+HostData = bin/archdatadir
+Data = bin/datadir
+Sysconf = bin/sysconfdir
+LibraryExecutables = bin/archdatadir/bin
+Plugins = bin/archdatadir/plugins
+Imports = bin/archdatadir/imports
+Qml2Imports = bin/archdatadir/qml
+Translations = bin/datadir/translations
+Documentation = bin/datadir/doc
+Examples = bin/datadir/examples""")
+        self.copy("*LICENSE*", src="qt5/", dst="licenses")
+        for module in self._submodules:
+            if module != 'qtbase' and not getattr(self.options, module):
+                tools.rmdir(os.path.join(self.package_folder, "licenses", module))
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
```

Comment:
Maybe I missed something, but how can we consume `qt` with CMake if cmake folder is removed and in the meantime no `package_info` (probably hard to maintain) is provided?

Reply:
for now, qmake is the only way to consume this package. I could implement `package_info` and defines the cmake targets, but the user will have to invoke moc and handle resources manually, because I don't think https://doc.qt.io/qt-5/cmake-command-reference.html can be reimplemented

Reply:
> or now, qmake is the only way to consume this package.

FWIW, [Qbs](http://qbs.io/) which queries qmake for metadata would work as well.

Reply:
I'm not sure if cmake described in https://doc.qt.io/qt-5/cmake-get-started.html could be emulated with conan components. Does someone with experience in components have an idea on this ?

Reply:
Cmake targets can be modeled I think. But for MOC, RCC features etc, I don't know.

Reply:
@ericLemanissier 
 I know that conan-center-index has a strict policy against cmake folder,but is it possible to add an option [default=False] to the recipe to package the cmake folder in order to consume your package with cmake?

Reply:
That would be a sneaky way. If there is enough people asking, I guess a more realistic approach would be to add qt as an exception to https://github.com/conan-io/hooks/blob/master/hooks/conan-center.py#L634

Reply:
or use `cpp_info.build_modules` maybe? https://github.com/conan-io/conan-center-index/issues/2151#issuecomment-659418239

Reply:
Hey,

package_info would be good here. Moc, rcc and uic work in cmake when you set Qt5_DIR in the consumer cmake file to coan-QT-rootfolder/lib/cmake/Qt5 and then find_package(COMPONENTS Core Widgets). After that you can use automoc and normal moc cmake commands as well. Then just use the targets created by conan. The only thing is windows. When you need a win32 exe then you should also link to the Qt5::Core target to get a qtmain.

Use this technique for years now, works pretty well so far.

Reply:
@tonka3000 you are saying that this works, even when `coan-QT-rootfolder/lib/cmake` is empty ?

Reply:
Sorry, missed the point that you dont ship the cmake folder from qt. Do you remove it because of a conan-center-index restriction?

Reply:
yes: https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#why-are-cmake-findconfig-files-and-pkg-config-files-not-packaged

Reply:
So the only way is to build a custom generator for the automoc, autouic, autores and qtmain (on windows). So every qt cmake function/macro needs to be built from the ground up. This make sense in most project but in Qt that's not a good idea in my opinion.

Reply:
@ericLemanissier Are build_modules allowed?

Reply:
AFAIK, build_modules  are allowed. if automoc et al are in standalone module files, we could save them in self.cpp_info.build_modules folder.

Reply:
So I've looked at this before. CMake implements AUTOGEN in it's own C++ source code. It only depends upon the executable targets `Qt5::moc`, `Qt5::rcc`, and `Qt5::uic` being appropriately defined.

As for the Qt's CMake macros referenced [here](https://doc.qt.io/qt-5/cmake-command-reference.html), they are defined separately in `Qt5{ComponentName}Macros.cmake`, where ComponentName is one of the components listed in the reference article. That said, they can depend on variables defined in the corresponding Config file. For instance, Qt5CoreMacros depends upon the cmake variables `Qt5Core_RCC_EXECUTABLE` and `Qt5Core_MOC_EXECUTABLE` being defined (a bit silly, since they are just defined to the corresponding executable targets). There are few enough of these though, that it would be reasonable to patch those variables out.

Given all of that, I think it would definitely be possible to model everything with Conan's components + build_modules.

Reply:
Oh, one thing I forgot. The way CMake picks up which version of Qt is being used for AUTOGEN is by looking at the values of the cmake variables `QT_VERSION_MAJOR`, `Qt5Core_VERSION_MAJOR`, `Qt6Core_VERSION_MAJOR`, and their corresponding minor versions. That would probably have to be injected via a build_module.

Reply:
don't hesitate to make PR targeting https://github.com/bincrafters/conan-center-index/tree/qt, I won't have much time available for this feature

Reply:
@ericLemanissier At the moment, conan doesn't handle executable components (see https://github.com/conan-io/conan/issues/7240 for discussion on build/host machine issues). I think the best solution for right now is to add qt as an exception for the cci hook, and include the cmake files provided by qt.

Reply:
Qt and cmake works fine as long as you set `CMAKE_PREFIX_PATH`.
I created my own Qt conan recipe and I set the followings stuff to `package_info`.
```
    def package_info(self):
        self.cpp_info.includedirs.append(os.path.join(prefix,"include"))
        self.cpp_info.libdirs.append(os.path.join(prefix,"lib"))
        self.cpp_info.frameworkdirs.append(os.path.join(prefix,"lib"))
        self.cpp_info.bindirs.append(os.path.join(prefix, "bin"))

        self.env_info.path.append(os.path.join(prefix, "bin"))
        self.env_info.CMAKE_PREFIX_PATH.append(prefix)
        self.env_info.QML2_IMPORT_PATH.append(os.path.join(prefix, "qml"))
        self.env_info.QT_PLUGIN_PATH.append(os.path.join(prefix, "plugins"))
```

Reply:
@bog-dan-ro, sure it works, but only because you don't remove cmake config files. Anyway CCI recipes should be consumable by any build system, so `package_info` must completly describes interface of the library to be able to create libs information for any conan generators.